tests/ipfwd/test_nhop_count.py

```diff
@@ -0,0 +1,324 @@
+            self.duthost.shell("rm {}".format(self.filename))
+            os.remove(self.filename)
+        except:
+            pass
+
+    def clean_up(self):
+        # delete static ARPs
+        self.arps_del()
+
+        # del IP address from the eth interface
+        ip_iface = "ip address del {} dev {}".format(self.if_addr, self.iface)
+        logging.info("IF ADDR DEL {}".format(ip_iface))
+        try:
+            self.asic.command(ip_iface)
+        except:
+            pass
+
+
+def get_crm_info(duthost, asic):
+    """
+    get CRM info
+    """
+    get_group_stats = ("{} COUNTERS_DB HMGET CRM:STATS"
+        " crm_stats_nexthop_group_used"
+        " crm_stats_nexthop_group_available").format(asic.sonic_db_cli)
+
+    result = duthost.command(get_group_stats)
+    pytest_assert(result["rc"] == 0, get_group_stats)
+
+    crm_info = {
+        "used": int(result["stdout_lines"][0]),
+        "available": int(result["stdout_lines"][1])
+    }
+
+    get_polling = '{} CONFIG_DB HMGET "CRM|Config" "polling_interval"'.format(
+        asic.sonic_db_cli
+    )
+    result = duthost.command(get_polling)
+    pytest_assert(result["rc"] == 0, get_polling)
+
+    crm_info.update({
+        "polling": int(result["stdout_lines"][0])
+    })
+
+    return crm_info
+
+
+# code from doc.python.org to generate combinations
+# This is used to create unique nexthop groups
+def combinations(iterable, r):
+    # combinations('ABCD', 2) --> AB AC AD BC BD CD
+    # combinations(range(4), 3) --> 012 013 023 123
+    pool = tuple(iterable)
+    n = len(pool)
+    if r > n:
+        return
+    indices = list(range(r))
+    yield tuple(pool[i] for i in indices)
+    while True:
+        for i in reversed(range(r)):
+            if indices[i] != i + n - r:
+                break
+        else:
+            return
+        indices[i] += 1
+        for j in range(i+1, r):
+            indices[j] = indices[j-1] + 1
+        yield tuple(pool[i] for i in indices)
+
+
+def loganalyzer_ignore_regex_list():
+    ignore = [
+        ".*Unaccounted_ROUTE_ENTRY_TABLE_entries.*"
+    ]
+    return ignore
+
+
+def test_nhop(request, duthost):
+    """
+    Test next hop group resource count. Steps:
+    - Add test IP address to an active IP interface
+    - Add static ARPs
+    - Create unique next hop groups
+    - Add IP route and nexthop
+    - check CRM resource
+    - clean up
+    - Verify no erros and crash
+    """
+    default_max_nhop_paths = 32
+    nhop_group_limit = 1024
+
+    asic = duthost.asic_instance()
+
+    # find out MAX NHOP group count supported on the platform
+    result = asic.run_redis_cmd(
+        argv = ["redis-cli", "-n", 6, "HGETALL", "SWITCH_CAPABILITY|switch"]
+    )
+    it = iter(result)
+    switch_capability = dict(zip(it, it))
+    max_nhop = switch_capability.get("MAX_NEXTHOP_GROUP_COUNT")
+    max_nhop = nhop_group_limit if max_nhop == None else int(max_nhop)
+    nhop_group_count = min(max_nhop, nhop_group_limit)
```

Comment:
if the DUT already has a few ECMP groups already programmed, do we need to subtract those from the `nhop_group_count`

Reply:
The intent of the test is to make sure if we program more than the advertised nexthops and there shouldn't be any crash. Considering this there is no need to worry about how many are programmed.

Reply:
I think in 201911 and 202012, if we program more the max adv. nexthops, SAI returns a `TABLE_FULL` failure and the device might crash.

Reply:
202012 seems to be fine. Will skip test for release 201911 and 201811.